packages/next-international/src/app/middleware/index.ts

```diff
@@ -1,42 +1,65 @@
 import { NextRequest, NextResponse } from 'next/server';
 
-const Response = NextResponse;
+import { LOCALE_COOKIE, LOCALE_HEADER } from '../../common/constants';
+import type { I18NMiddlewareConfig } from '../../types';
+
+const DEFAULT_STRATEGY = 'redirect';
 
 export function createI18nMiddleware<Locales extends readonly string[]>(
   locales: Locales,
   defaultLocale: Locales[number],
+  config?: I18NMiddlewareConfig,
 ) {
   return function I18nMiddleware(request: NextRequest) {
-    const pathname = request.nextUrl.pathname;
-    const pathnameIsMissingLocale = locales.every(
-      locale => !pathname.startsWith(`/${locale}/`) && pathname !== `/${locale}`,
-    );
+    const requestUrl = request.nextUrl.clone();
 
-    if (pathnameIsMissingLocale && pathname !== '/favicon.ico') {
-      let locale = request.cookies.get('Next-Locale')?.value;
+    const locale = localeFromRequest(request, locales, config?.contentNegotiator) ?? defaultLocale;
 
-      if (!locale) {
-        const acceptLanguage = request.headers.get('Accept-Language');
-        const acceptLanguageLocale = acceptLanguage?.split(',')?.[0]?.split('-')?.[0] ?? defaultLocale;
+    if (noLocalePrefix(locales, requestUrl.pathname)) {
+      const mappedUrl = requestUrl.clone();
+      mappedUrl.pathname = `/${locale}${mappedUrl.pathname}`;
 
-        locale = locales.includes(acceptLanguageLocale) ? acceptLanguageLocale : defaultLocale;
+      const strategy = config?.urlMappingStrategy ?? DEFAULT_STRATEGY;
+      if (strategy === 'redirect') {
+        const response = NextResponse.redirect(mappedUrl);
+        return addLocaleToResponse(response, locale);
+      } else if (strategy === 'rewrite') {
+        const response = NextResponse.rewrite(mappedUrl);
+        return addLocaleToResponse(response, locale);
+      } else {
+        throw new Error(`Invalid urlMappingStrategy: ${strategy}`);
```

Comment:
I think we should fallback to `redirect` and `console.warn` instead of throwing an error in the middleware, which can break production